lib/rubotnik/message_dispatch.rb

```diff
@@ -56,8 +56,8 @@ def handle_command(to, reply_with)
       if reply_with.empty?
         puts "Command #{to} is executed for user #{@user.id}"
         execute(to)
-        @user.reset_command
-        puts "Command is reset for user #{@user.id}"
+        # WHY?? @user.reset_command
```

Comment:
This line broke my `expect..to receive` test once and it made me think why exactly do we need to reset the current command? Is it a mistake?

Reply:
So, here was my logic here: the `default` is only invoked at the top level of a conversation when his text is not matched against any bindings. If a user is already inside the command chain — it will never be triggered, it's the responsibility of commands to handle possible further input or call `stop_thread`. But if the user is somehow back to "top-level" and the writer of the code forgot to call `stop_thread` in a previous command, it will lead to weird behavior, as the user will still have a command and `MessageDispatch` will always try to execute it, so will be no way to get out of the loop. Or am I missing something?

Reply:
I actually meant to put this comment in the corresponding line in `postback_dispatch` as the state is already cleared on line #32 (https://github.com/progapandist/rubotnik/blob/master/lib/rubotnik/postback_dispatch.rb#L32).

Your explanation makes sense, however this line may also break a following case:
1. In bindings you have `bind "one", to: :step1`
2. `step1` method looks as follows:
```
def step1
  say text, quick_replies: [
    ['A', 'B']
  ]
  next_command :step2
end
```
3. Current behavior will cause that `step2` will not be executed at all. The only way to handle this currently is to pass a block to `bind`:
```
bind "one" do
  step1
end
```

Maybe we could change the line to the following?
```
# Reset the top level command if the current command hadn't changed
@user.reset_command if @user.current_command == to
```

Reply:
Actually, if a user has a command, `bind` logic will not be executed at all, as it only concerns "top-level" bindings, see https://github.com/progapandist/rubotnik/blob/7a34d9e62792dfbb7a9b1fdf214a684e34b1b2ac/lib/rubotnik/message_dispatch.rb#L20. So, effectively, `reply_with` can only be used at top level and it starts the thread (make Dispatch wait for the input for user that will be used for next assigned command, not at top level). If the top-level binging exists without `reply_with` specified, we can effectively reset the command, because it's just a "one off" thing: get the message, reply to the message, forget about it — new input will be handled at top level again. In `PostbackDispatch` it's a bit different, we can clear state each time, as there are no chained postbacks, it's just a trigger action.

Reply:
If you agree with my explanation (I agree very well that my logic may be flawed), then just remove comments and I'll merge :)

Reply:
@progapandist 

> If the top-level binging exists without reply_with specified, we can effectively reset the command, because it's just a "one off" thing: get the message, reply to the message, forget about it

That's not exactly true - check out the example in my previous comment. 
What I'd expect from this code:
```
module Commands
	def step1
	  say 'xxx', quick_replies: [
	    ['A', 'B']
	  ]
	  next_command :step2
	end
end	

bind "one", to: :step1
```

is to work the same as: `bind "one", to: :step2, reply_with: { text: xxx', quick_replies: ['A', 'B'] }`, but currently it'll just say 'xxx' and return to the top level.

Sorry if I didn't explain it clearly, but hope you understand this case.

Reply:
Thanks for expanding, I see your concern. For now it's expected behaviour though, we even say in the docs that you can only open a thread from top-level by using `reply_with`. Your use-case is completely justified though, so I agree that the current API is not clear. However, I'd like to take some time to rethink it and propose other solution, and for that it would be good to have your tests :) Let's keep it like it is for now, merge and bump the version that has tests and a bugfix for buttons, then I'll dive in the code and do the general housekeeping.